From 103db0b077193046ef44edd0984e65352b657e3a Mon Sep 17 00:00:00 2001
From: Eric Ruei <e-ruei1@ti.com>
Date: Fri, 27 Jan 2017 10:09:42 -0500
Subject: [PATCH] HACK: quota/client_usage_tracker: fix a mysterious crash at
 std::map

There is a mysterious segmentation fault when the first node of UsageMap or
HostUsageMap is created at function ClientUsageTracker::UpdateUsageCache.
There is no stack overflow and/or memory corruption observed and
the similar code works at some other modules.

It seems to be a compiler related bug and the temporary workaround is to
use uint64 instead of int64 so that the original functionality is still
perserved.

Signed-off-by: Eric Ruei <e-ruei1@ti.com>
---
 src/3rdparty/chromium/storage/browser/quota/client_usage_tracker.cc | 8 ++++----
 src/3rdparty/chromium/storage/browser/quota/client_usage_tracker.h  | 2 +-
 2 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/src/3rdparty/chromium/storage/browser/quota/client_usage_tracker.cc b/src/3rdparty/chromium/storage/browser/quota/client_usage_tracker.cc
index 968b351..57d4178 100644
--- a/src/3rdparty/chromium/storage/browser/quota/client_usage_tracker.cc
+++ b/src/3rdparty/chromium/storage/browser/quota/client_usage_tracker.cc
@@ -142,7 +142,7 @@ void ClientUsageTracker::UpdateUsageCache(
     if (!IsUsageCacheEnabledForOrigin(origin))
       return;
 
-    cached_usage_by_host_[host][origin] += delta;
+    cached_usage_by_host_[host][origin] += (uint64)delta;
     if (IsStorageUnlimited(origin))
       global_unlimited_usage_ += delta;
     else
@@ -348,9 +348,9 @@ void ClientUsageTracker::AddCachedOrigin(
   DCHECK(IsUsageCacheEnabledForOrigin(origin));
 
   std::string host = net::GetHostOrSpecFromURL(origin);
-  int64* usage = &cached_usage_by_host_[host][origin];
-  int64 delta = new_usage - *usage;
-  *usage = new_usage;
+  uint64* usage = &cached_usage_by_host_[host][origin];
+  int64 delta = new_usage - (int64)*usage;
+  *usage = (uint64)new_usage;
   if (delta) {
     if (IsStorageUnlimited(origin))
       global_unlimited_usage_ += delta;
diff --git a/src/3rdparty/chromium/storage/browser/quota/client_usage_tracker.h b/src/3rdparty/chromium/storage/browser/quota/client_usage_tracker.h
index 1c07dc8..e5de13a 100644
--- a/src/3rdparty/chromium/storage/browser/quota/client_usage_tracker.h
+++ b/src/3rdparty/chromium/storage/browser/quota/client_usage_tracker.h
@@ -62,7 +62,7 @@ class ClientUsageTracker : public SpecialStoragePolicy::Observer,
       HostUsageAccumulatorMap;
 
   typedef std::set<std::string> HostSet;
-  typedef std::map<GURL, int64> UsageMap;
+  typedef std::map<GURL, uint64> UsageMap;
   typedef std::map<std::string, UsageMap> HostUsageMap;
 
   struct AccumulateInfo {
-- 
1.9.1

